peer/peerdiscovery.go

```diff
@@ -0,0 +1,147 @@
+package peer
+
+import (
+	"context"
+	"encoding/binary"
+	"fmt"
+	"github.com/renproject/surge"
+	"time"
+
+	"github.com/renproject/aw/transport"
+	"github.com/renproject/aw/wire"
+	"github.com/renproject/id"
+
+	"go.uber.org/zap"
+)
+
+type DiscoveryClient struct {
+	opts DiscoveryOptions
+
+	transport *transport.Transport
+}
+
+func NewDiscoveryClient(opts DiscoveryOptions, transport *transport.Transport) *DiscoveryClient {
+	return &DiscoveryClient{
+		opts: opts,
+		transport: transport,
+	}
+}
+
+func (dc *DiscoveryClient) DiscoverPeers(ctx context.Context) {
+	var pingData [2]byte
+	binary.LittleEndian.PutUint16(pingData[:], dc.transport.Port())
+
+	msg := wire.Msg{
+		Version: wire.MsgVersion1,
+		Type:    wire.MsgTypePing,
+		Data:    pingData[:],
+	}
+
+	ticker := time.NewTicker(dc.opts.PingTimePeriod)
+	defer ticker.Stop()
+
+	alpha := dc.opts.Alpha
+	sendDuration := dc.opts.PingTimePeriod/time.Duration(alpha)
+	outer:
+	for {
+		for _, sig := range dc.transport.Table().Peers(alpha) {
+			innerCtx, cancel := context.WithTimeout(ctx, sendDuration)
+			msg.To = id.Hash(sig)
+			err := dc.transport.Send(innerCtx, sig, msg)
+			cancel()
```

Comment:
```suggestion
			err := func() error {
				innerCtx, innerCancel := context.WithTimeout(ctx, sendDuration)
				defer innerCancel()
				msg.To = id.Hash(sig)
				return dc.transport.Send(innerCtx, sig, msg)
			}()
```
You should always try and use the `defer` statement, especially when making calls to functions, because it will ensure that the deferred statement is _always_ run (even if one of the function calls panics).